mantidproject/mantid
# Improve sampling of scatter points in absorption correction simulation

**Description of work.**

Part of the Monte Carlo calculation for absorption corrections involves generating random scattering points in either the sample or the environment. The sampling of these points didn't properly\fairly take into account the volume of each component. The sampling is now performed according to the volume of each part of the sample\environment that intersects the beam profile

Following changes have been made:

a) the function MeshObject::generatePointInObject has been modified so that when it doesn't find a point it returns false rather than raising an exception. This better supports usage with maxAttempts=1 where the calling code cycles through the environment components checking if a random point in the "active region" lies within each component. This involved a change in the function signature that has been made in IObject and all of the child classes (MeshObject, CSGObject, MeshObject2D, Container)

b) new function MCInteractionVolume::generatePoint that randomly generates a scatter point across the sample and environment components. This replaces a similar function that was in SampleEnvironment (that one only considered the env components and didn't include the sample). This leaves the SampleEnvironment class a bit light on functionality. Possible it could be retired entirely and just replaced by a vector of IObject items.

c) added some logging to the simulation to show where the scatter points occurred. These show that for Pearl around 6% of the scatter points are in the sample which is less than the 50% assumption previously in the code

d) small change in CSGObject::generatePointInObject to only try the fallback method if there are any attempts left. The fallback method isn't v helpful for correct sampling of scatter points in the sample\environment cpts because it always finds a point in the CSGObject if the object's bounding box is inside the active region, irrespective of how much of the active region the object occupies. This change modifies the results by a small amount and so required a change to the tests (MonteCarloAbsorptionTest.h and AbsorptionShapesTest.py). As well as updating the expected values I have also reduced the level of precision

**To test:**

Unzip [AbsorptionCorrectionSampling.zip](https://github.com/mantidproject/mantid/files/4064906/AbsorptionCorrectionSampling.zip) and place all the files in a directory that is specified Data Search Directories in "Manage User Directories". Run the .py script which does two things:

a) creates a workspace called "sample_env_mesh" containing
a mesh representation of the Pearl sample\environment with a single detector loaded. The script loads a mesh description of each environment component from a .stl file

b) runs the MonteCarloAbsorption algorithm on the sample_env_mesh workspace
to calculate an absorption correction and outputs the results into workspace
"sample_env_mesh_abscorr"

Having done this you can do these two things to test:
- Plot the single spectrum in the sample_env_mesh_abscorr workspace (which is the absorption correction for the single detector) and check there's a valid absorption plot (~0.05-0.1 and decreasing with wavelength)
- Note that the debug messages show the scatter points have been sampled according
to the volume intersecting the beam profile


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
